File: TwoPointers/15. 3Sum/3Sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
    let result = [];
    if (nums.length < 3) return result;

    nums.sort((a, b) => a - b);

    for (let i = 0; i < nums.length; i++) {
        // if the smallest number is > 0, forget about it
        if (nums[i] > 0) break;
        // if the current nums[i] is the same as nums[i - 1]      
        // means there is a combination added to the array, skip
        if (i > 0 && nums[i - 1] === nums[i]) {
            continue;
        }

        // two sum II - two pointers     
        let l = i + 1;
        let r = nums.length - 1;

        while (l < r) {
            let sum = nums[i] + nums[l] + nums[r];
            if (sum === 0) {
                result.push([nums[i], nums[l], nums[r]]);

                // move pointers ahead   

                // 1. for case [-2, -2, 0, 0, 2, 2] to avoid repetition
                while (l < r && nums[l] === nums[l + 1]) {
                    l++;
                }
                while (l < r && nums[r] === nums[r - 1]) {
                    r--;
                }
                // 2. can't break like two sum II
                // as two sum II only has exactly one solution
                // this logic goes after the two while loop above
                // for case [-2, 0, 1, 1, 2]
                // at -2, if l and r moves then will be at 1 and 1
                // now nums[r] === nums[r - 1], then r will move
                // and now r === l will break out the loop
                l++;
                r--;

            }
            if (sum > 0) {
                r--;
            }
            if (sum < 0) {
                l++;
            }
        }
    }

    return result;
};

File: TwoPointers/15. 3Sum/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        if len(nums) < 3:
            return result
        
        # sort
        nums.sort()
        
        # for index, value in enumerate(arr)
        for i, num in enumerate(nums):
            if num > 0: 
                break
            if i > 0 and num == nums[i - 1]:
                continue
            
            l = i + 1
            r = len(nums) - 1
            
            while l < r:
                sum = num + nums[l] + nums[r]
                if sum > 0:
                    r -= 1
                if sum < 0:
                    l += 1
                if sum == 0:
                    result.append([num, nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    l += 1
                    r -= 1
                    
        return result
            